In [321]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

# Data Preparation

### Loading Both Datasets

df1 = fear_greed_index.csv  
df2 = historical_data.csv

In [322]:
df1 = pd.read_csv('fear_greed_index.csv')
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2644 entries, 0 to 2643
Data columns (total 4 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   timestamp       2644 non-null   int64 
 1   value           2644 non-null   int64 
 2   classification  2644 non-null   object
 3   date            2644 non-null   object
dtypes: int64(2), object(2)
memory usage: 82.8+ KB


In [323]:
print(df1.duplicated().sum())

0


No missing values found in df1  
No duplicates found in df1

In [324]:
df2 = pd.read_csv('historical_data.csv')
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 211224 entries, 0 to 211223
Data columns (total 16 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   Account           211224 non-null  object 
 1   Coin              211224 non-null  object 
 2   Execution Price   211224 non-null  float64
 3   Size Tokens       211224 non-null  float64
 4   Size USD          211224 non-null  float64
 5   Side              211224 non-null  object 
 6   Timestamp IST     211224 non-null  object 
 7   Start Position    211224 non-null  float64
 8   Direction         211224 non-null  object 
 9   Closed PnL        211224 non-null  float64
 10  Transaction Hash  211224 non-null  object 
 11  Order ID          211224 non-null  int64  
 12  Crossed           211224 non-null  bool   
 13  Fee               211224 non-null  float64
 14  Trade ID          211224 non-null  float64
 15  Timestamp         211224 non-null  float64
dtypes: bool(1), float64(

In [325]:
print(df2.duplicated().sum())

0


No missing values found in df2  
No duplicates found in df2

In [326]:
df2.describe()

Execution Price   Size Tokens      Size USD  Start Position  \
count    211224.000000  2.112240e+05  2.112240e+05    2.112240e+05   
mean      11414.723350  4.623365e+03  5.639451e+03   -2.994625e+04   
std       29447.654868  1.042729e+05  3.657514e+04    6.738074e+05   
min           0.000005  8.740000e-07  0.000000e+00   -1.433463e+07   
25%           4.854700  2.940000e+00  1.937900e+02   -3.762311e+02   
50%          18.280000  3.200000e+01  5.970450e+02    8.472793e+01   
75%         101.580000  1.879025e+02  2.058960e+03    9.337278e+03   
max      109004.000000  1.582244e+07  3.921431e+06    3.050948e+07   

          Closed PnL      Order ID            Fee      Trade ID     Timestamp  
count  211224.000000  2.112240e+05  211224.000000  2.112240e+05  2.112240e+05  
mean       48.749001  6.965388e+10       1.163967  5.628549e+14  1.737744e+12  
std       919.164828  1.835753e+10       6.758854  3.257565e+14  8.689920e+09  
min   -117990.104100  1.732711e+08      -1.175712  0.000000e+00  1.680000e+12  
25%         0.000000  5.983853e+10       0.016121  2.810000e+14  1.740000e+12  
50%         0.000000  7.442939e+10       0.089578  5.620000e+14  1.740000e+12  
75%         5.792797  8.335543e+10       0.393811  8.460000e+14  1.740000e+12  
max    135329.090100  9.014923e+10     837.471593  1.130000e+15  1.750000e+12

### Converting Timestamps and aligning datasets by date

In [327]:
df1['date'] = pd.to_datetime(df1['date']).dt.normalize()
df2['date'] = pd.to_datetime(df2['Timestamp IST'], format="%d-%m-%Y %H:%M").dt.normalize()

In [328]:
df_merged = df2.merge(df1[['date','classification']], on='date', how='left')
df_merged.head()

Account  Coin  Execution Price  \
0  0xae5eacaf9c6b9111fd53034a602c192a04e082ed  @107           7.9769   
1  0xae5eacaf9c6b9111fd53034a602c192a04e082ed  @107           7.9800   
2  0xae5eacaf9c6b9111fd53034a602c192a04e082ed  @107           7.9855   
3  0xae5eacaf9c6b9111fd53034a602c192a04e082ed  @107           7.9874   
4  0xae5eacaf9c6b9111fd53034a602c192a04e082ed  @107           7.9894   

   Size Tokens  Size USD Side     Timestamp IST  Start Position Direction  \
0       986.87   7872.16  BUY  02-12-2024 22:50        0.000000       Buy   
1        16.00    127.68  BUY  02-12-2024 22:50      986.524596       Buy   
2       144.09   1150.63  BUY  02-12-2024 22:50     1002.518996       Buy   
3       142.98   1142.04  BUY  02-12-2024 22:50     1146.558564       Buy   
4         8.73     69.75  BUY  02-12-2024 22:50     1289.488521       Buy   

   Closed PnL                                   Transaction Hash     Order ID  \
0         0.0  0xec09451986a1874e3a980418412fcd0201f500c95bac...  52017706630   
1         0.0  0xec09451986a1874e3a980418412fcd0201f500c95bac...  52017706630   
2         0.0  0xec09451986a1874e3a980418412fcd0201f500c95bac...  52017706630   
3         0.0  0xec09451986a1874e3a980418412fcd0201f500c95bac...  52017706630   
4         0.0  0xec09451986a1874e3a980418412fcd0201f500c95bac...  52017706630   

   Crossed       Fee      Trade ID     Timestamp       date classification  
0     True  0.345404  8.950000e+14  1.730000e+12 2024-12-02  Extreme Greed  
1     True  0.005600  4.430000e+14  1.730000e+12 2024-12-02  Extreme Greed  
2     True  0.050431  6.600000e+14  1.730000e+12 2024-12-02  Extreme Greed  
3     True  0.050043  1.080000e+15  1.730000e+12 2024-12-02  Extreme Greed  
4     True  0.003055  1.050000e+15  1.730000e+12 2024-12-02  Extreme Greed

### Daily PnL and Number of Trades of each Trader

In [330]:
# Daily pnl per trader
daily_pnl = df_merged.groupby(['Account','date','classification'])['Closed PnL'].sum().reset_index(name='total_pnl')

# Number of trades per trader
daily_trades = df_merged.groupby(['Account','date']).size().reset_index(name='num_trades')

daily_trader_data = daily_pnl.merge(daily_trades, on=['date','Account'])
daily_trader_data

Account       date classification  \
0     0x083384f897ee0f19899168e3b1bec365f52a9012 2024-11-11  Extreme Greed   
1     0x083384f897ee0f19899168e3b1bec365f52a9012 2024-11-17  Extreme Greed   
2     0x083384f897ee0f19899168e3b1bec365f52a9012 2024-11-18  Extreme Greed   
3     0x083384f897ee0f19899168e3b1bec365f52a9012 2024-11-22  Extreme Greed   
4     0x083384f897ee0f19899168e3b1bec365f52a9012 2024-11-26  Extreme Greed   
...                                          ...        ...            ...   
2335  0xbee1707d6b44d4d52bfe19e41f8a828645437aab 2025-04-27          Greed   
2336  0xbee1707d6b44d4d52bfe19e41f8a828645437aab 2025-04-28        Neutral   
2337  0xbee1707d6b44d4d52bfe19e41f8a828645437aab 2025-04-29          Greed   
2338  0xbee1707d6b44d4d52bfe19e41f8a828645437aab 2025-04-30          Greed   
2339  0xbee1707d6b44d4d52bfe19e41f8a828645437aab 2025-05-01        Neutral   

         total_pnl  num_trades  
0         0.000000         177  
1         0.000000          68  
2         0.000000          40  
3    -21227.000000          12  
4      1603.100000          27  
...            ...         ...  
2335   1709.194807          82  
2336   4008.588908         430  
2337   8561.771838         902  
2338   2520.773814          75  
2339   1364.022527         131  

[2340 rows x 5 columns]

### Historical Information of each Account

In [331]:
# Determining Win Rate
closed_trades = df_merged[df_merged['Closed PnL'] != 0]
closed_trades['win'] = closed_trades['Closed PnL'] > 0
win_rate = closed_trades.groupby('Account')['win'].mean().sort_values().reset_index(name='win_rate')

# Other trader aggregates
trader_summary = df_merged.groupby('Account').agg(total_pnl = ('Closed PnL', 'sum'), avg_pnl = ('Closed PnL', 'mean'),pnl_std = ('Closed PnL', 'std'), num_trades = ('Closed PnL', 'count'), avg_trade_size = ('Size Tokens', 'median')).reset_index()

# Merging into single trader_info DataFrame
trader_summary = win_rate.merge(trader_summary, on='Account')

trader_summary.head()

C:\Users\divya\AppData\Local\Temp\ipykernel_22036\3453849233.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  closed_trades['win'] = closed_trades['Closed PnL'] > 0


Account  win_rate      total_pnl  \
0  0x8477e447846c758f5a675856001ea72298fd9cb5  0.599024   43917.008976   
1  0x3998f134d6aaa2b6a5f723806d00fd2bbbbce891  0.650877  -31203.599986   
2  0x39cef799f8b69da1995852eea189df24eb5cae3c  0.665535   14456.919336   
3  0x271b280974205ca63b716753467d5a371de622ab  0.715619  -70436.191318   
4  0xbd5fead7180a9c139fa51a103cb6a2ce86ddb5c3  0.734919  220519.061853   

     avg_pnl      pnl_std  num_trades  avg_trade_size  
0   2.928191    68.613041       14998        299.0000  
1 -38.286626   558.778753         815        196.7200  
2   4.028119   730.495484        3589        358.6000  
3 -18.492043   542.182659        3809          1.8178  
4  83.498320  1530.525033        2641        100.0000

### Trader Segmentation

#### Frequent vs Infrequent

In [333]:
median_trades = trader_summary['num_trades'].median()

trader_summary['frequency_segment'] = trader_summary['num_trades'].apply(lambda x: 'frequent' if  x > median_trades else 'infrequent')

#### Consistent vs Inconsistent

In [334]:
mean_winrate = trader_summary['win_rate'].mean()

trader_summary['consistency_segment'] = trader_summary['win_rate'].apply(lambda x: 'consistent' if  x > mean_winrate else 'inconsistent')

#### High Risk vs Low Risk (Leverage Proxy)

In [335]:
median_size = trader_summary['avg_trade_size'].median()

trader_summary['risk_segment'] = trader_summary['avg_trade_size'].apply(lambda x: 'High Risk' if x > median_size else 'Low Risk')

In [336]:
trader_summary

Account  win_rate     total_pnl  \
0   0x8477e447846c758f5a675856001ea72298fd9cb5  0.599024  4.391701e+04   
1   0x3998f134d6aaa2b6a5f723806d00fd2bbbbce891  0.650877 -3.120360e+04   
2   0x39cef799f8b69da1995852eea189df24eb5cae3c  0.665535  1.445692e+04   
3   0x271b280974205ca63b716753467d5a371de622ab  0.715619 -7.043619e+04   
4   0xbd5fead7180a9c139fa51a103cb6a2ce86ddb5c3  0.734919  2.205191e+05   
5   0x72743ae2822edd658c0c50608fd7c5c501b2afbd  0.746269  4.293556e+05   
6   0xaf40fdc468c30116bd3307bcbf4a451a7ebf1deb  0.747826  2.175883e+04   
7   0x8170715b3b381dffb7062c0298972d4727a0a63b  0.752243 -1.676211e+05   
8   0xbee1707d6b44d4d52bfe19e41f8a828645437aab  0.763070  8.360806e+05   
9   0x23e7a7f8d14b550961925fbfdaa92f5d195ba5bd  0.773644  4.788532e+04   
10  0x72c6a4624e1dffa724e6d00d64ceae698af892a0  0.776596  4.030115e+05   
11  0xb1231a4a2dd02f2276fa3c5e2a2f3436e6bfed23  0.791050  2.143383e+06   
12  0x083384f897ee0f19899168e3b1bec365f52a9012  0.792725  1.600230e+06   
13  0xb899e522b5715391ae1d4f137653e7906c5e2115  0.803484  2.248850e+04   
14  0x2c229d22b100a7beb69122eed721cee9b24011dd  0.817476  1.686580e+05   
15  0x47add9a56df66b524d5e2c1993a43cde53b6ed85  0.822004  1.033437e+05   
16  0x28736f43f1e871e6aa8b1148d38d4994275d72c4  0.862079  1.324648e+05   
17  0x8381e6d82f1affd39a336e143e081ef7620a3b7f  0.883754  6.551366e+04   
18  0x513b8629fe877bb581bf244e326a047b249c4ff1  0.895476  8.404226e+05   
19  0x75f7eeb85dc639d5e99c78f95393aa9a5f1170d4  0.926328  3.790954e+05   
20  0x4f93fead39b70a1824f981a54d4e55b278e9f760  0.926755  3.089759e+05   
21  0xa520ded057a32086c40e7dd6ed4eb8efb82c00e0  0.937255  7.284648e+04   
22  0x7f4f299f74eec87806a830e3caa9afa5f2b9db8f  0.945040  1.490044e+04   
23  0x4acb90e786d897ecffb614dc822eb231b4ffb9f4  0.948500  6.777471e+05   
24  0xa0feb3725a9335f49874d7cd8eaad6be45b27416  0.970329  1.063029e+05   
25  0x92f17e8d81a944691c10e753af1b1baae1a2cd0d  0.976457  1.265789e+05   
26  0x6d6a4b953f202f8df5bed40692e7fd865318264a  0.988263  1.087312e+05   
27  0xbaaaf6571ab7d571043ff1e313a9609a10637864  0.991197  9.401638e+05   
28  0x3f9a0aadc7f04a7c9d75dc1b5a6ddd6e36486cf6  1.000000  5.349625e+04   
29  0x420ab45e0bd8863569a5efbb9c05d91f40624641  1.000000  1.995056e+05   
30  0xae5eacaf9c6b9111fd53034a602c192a04e082ed  1.000000  6.784562e+04   
31  0x430f09841d65beb3f27765503d0f850b8bce7713  1.000000  4.165419e+05   

       avg_pnl      pnl_std  num_trades  avg_trade_size frequency_segment  \
0     2.928191    68.613041       14998      299.000000          frequent   
1   -38.286626   558.778753         815      196.720000        infrequent   
2     4.028119   730.495484        3589      358.600000        infrequent   
3   -18.492043   542.182659        3809        1.817800          frequent   
4    83.498320  1530.525033        2641      100.000000        infrequent   
5   270.034947  2525.240437        1590       10.000000        infrequent   
6    40.746876   288.266124         534        0.012430        infrequent   
7   -36.431455  1647.778984        4601      238.200000          frequent   
8    20.806305   264.123880       40184       33.170000          frequent   
9     6.577654   134.073473        7280       24.985000          frequent   
10  281.826227  1591.523958        1430       50.000000        infrequent   
11  145.481748  1374.026402       14733       25.550000          frequent   
12  419.127768  4692.452329        3818        1.470000          frequent   
13    4.648305   332.083405        4838        7.740000          frequent   
14   52.071011   255.829091        3239      998.300000        infrequent   
15   12.130968    88.292842        8519       40.200000          frequent   
16    9.951530   113.269353       13311       24.900000          frequent   
17   34.282396   325.755610        1911       30.240000        infrequent   
18   68.684419   639.402444       12236        0.158685          frequent   
19   38.319560   172.520773        9893     1219.000000 

# Predictive Model #1

##### This model is based on trader's long run characteristics. 
##### To predict whether a trader will make a profit on a particular day or not, we only need to provide   
##### (1) trader characteristics (frequent/infrequent, consistent/inconsistent, High Risk/Low Risk) and  
##### (2) market sentiment on that day (Extreme Fear/Fear/Neutral/Greed/Extreme Greed)

### Features and Labels

In [ ]:
features = ['classification','frequency_segment','consistency_segment','risk_segment']
labels = ['profit?']

columns_unprepared = ['classification','frequency_segment','consistency_segment','risk_segment','total_pnl_x']

main_data = daily_trader_data.merge(trader_summary, on='Account', how='left')[columns_unprepared]
main_data['profit?'] = main_data['total_pnl_x'].apply(lambda x: 'y' if x>0 else 'n')

X = main_data[features]
y = main_data[labels]

### Handling Categorical Variables

In [338]:
from sklearn.preprocessing import OrdinalEncoder, LabelEncoder

oe = OrdinalEncoder()
X_encoded = oe.fit_transform(X)

le = LabelEncoder()
y_encoded = le.fit_transform(y)

c:\Users\divya\anaconda3\Lib\site-packages\sklearn\preprocessing\_label.py:114: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


### Model Selection

In [339]:
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import cross_val_score

In [340]:
# Testing Logistic Regression Model
lr = LogisticRegression()
lr_scores = cross_val_score(estimator=lr, X=X_encoded, y=y_encoded, cv=10, scoring='roc_auc')
print(f"The mean of Logistic Regression ROC-AUC scores is {lr_scores.mean()}")

The mean of Logistic Regression ROC-AUC scores is 0.6252341870889007


In [341]:
# Testing Decision Tree Classifier Model
cdt = DecisionTreeClassifier(random_state=42)
cdt_scores = cross_val_score(estimator=cdt, X=X_encoded, y=y_encoded, cv=10, scoring='roc_auc')
print(f"The mean of Decision Tree Classifier ROC-AUC scores is {cdt_scores.mean()}")

The mean of Decision Tree Classifier ROC-AUC scores is 0.5108675693191349


In [342]:
# Testing Random Forest Classifier Model
crf = RandomForestClassifier(random_state=42)
crf_scores = cross_val_score(estimator=crf, X=X_encoded, y=y_encoded, cv=10, scoring='roc_auc')
print(f"The mean of Random Forest Classifier ROC-AUC scores is {crf_scores.mean()}")

The mean of Random Forest Classifier ROC-AUC scores is 0.5094218882354047


#### Logistic Regression has highest mean ROC-AUC score (~0.62)

### Data Partition

In [343]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(X_encoded, y_encoded, test_size=0.2, random_state=42)

### Training and Prediction

In [344]:
lr = lr.fit(x_train, y_train)
predictions = lr.predict(x_test)

from sklearn.metrics import roc_auc_score
score = roc_auc_score(y_test, predictions)
print(score)

0.5919367423449484


##### The ROC-AUC Score of our model is ~0.59